src/runtime-rs/crates/hypervisor/src/device/topology.rs

```diff
@@ -0,0 +1,240 @@
+-----------       -----------         -----------                -----------
+| PCI Dev |       | PCI Dev |         | PCI Dev |                | PCI Dev |
+-----------       -----------         -----------                -----------
+
+Plug only PCI Express devices into PCI Express Ports.
+pcie.0 bus (Root Complex)
+----------------------------------------------------------------------------
+|    |    |    |    |    |    |    |    |    |    |    |    |    |    | .. |
+------|----------------|--------------------------------------|-------------
+      |                |                                      |
+      V                V                                      V
+ -------------     -------------                        -------------
+ | Root Port |     | Root Port |                        | Root Port |
+ -------------     -------------                        -------------
+     |                                                       |
+     |                              -------------------------|-----------------------
+------------                        |                -----------------              |
+| PCIe Dev |                        |   PCI Express  | Upstream Port |              |
+------------                        |     Switch     -----------------              |
+                                    |                 |            |                |
+                                    |   -------------------    -------------------  |
+                                    |   | Downstream Port |    | Downstream Port |  |
+                                    |   -------------------    -------------------  |
+                                    -------------|-----------------------|-----------
+                                            ------------
+                                            | PCIe Dev |
+                                            ------------
+*/
+
+use std::collections::HashMap;
+
+use crate::device::pci_path::PciSlot;
+use anyhow::Result;
+use kata_types::config::hypervisor::TopologyConfigInfo;
+
+use super::pci_path::PciPath;
+use crate::HYPERVISOR_DRAGONBALL;
+
+pub const PCIE_ROOT_BUS_SLOTS_CAPACITY: u32 = 32;
+pub const PCIE_TOPOLOGY_SLOTS_CAPACITY: u32 = 32;
+const DEFAULT_PCIE_ROOT_BUS: &str = "pcie.0";
+const DEFAULT_PCIE_ROOT_BUS_ADDRESS: &str = "0000:00";
+
+pub trait PCIeDevice: Send + Sync {
+    fn device_type(&self) -> PCIeDevType;
+    fn device_id(&self) -> &str;
+}
+
+// PCI-PCI Bridge
+pub struct PCIBridge {
+    _pci_endpoints: HashMap<u32, PCIeEndpoint>,
+}
+
+#[derive(Default)]
+pub struct PCIePCIBridge {
+    pub root_port: PCIeRootPort,
+    _bridge_devices: HashMap<String, PCIeEndpoint>,
+}
+
+// PCIe Root Complex
+#[derive(Clone, Debug, Default)]
+pub struct PCIeRootComplex {
+    pub root_bus: String,
+    pub root_bus_address: String,
+    pub root_bus_devices: HashMap<String, PCIeEndpoint>,
+}
+
+// PCIe Root Port
+#[derive(Clone, Debug, Default)]
+pub struct PCIeRootPort {
```

Comment:
Is it going to be possible to set mem-reserver and pref-64-reserve for the root port? Does dragonball has such settings for the root port? The GPU has "huge" BARs and they need proper IOMMU memory setup. We cannot work with the default 4k window range.

Reply:
Thx @zvonkok 
1. Setting mem-reserve and pref-64-reserve is to be supported, and based on my analysis of kata-runtime, it seems that only qemu needs to set the resource reservation, am I right?
2. AFAIK, Dragonball is a lightweight VMM and now only supports PCIe/PCI devices directly attaching onto the root bus. IOW, it has no such setting for the root port now.  [Discussion about it can be seen at slack ](https://katacontainers.slack.com/archives/C02SN6U9P1D/p1698844006082159)

Reply:
You're right only QEMU needs that for now. We should keep in mind that some other HV may need this as well. If we can make it generic this would be great. 
Understood so this PR is to get Dragonball working.

Reply:
Yes, It's designed to cover all VMMs, not dedicated for some VMM, but now it's the first step to support dragonball and keep compatibility with CLH. And the next, I'll plan to support Qemu.